docs/aspnetcore-facility.md

```diff
@@ -10,38 +10,78 @@ The ASP.NET Core facility provides Castle Windsor integration using a custom act
-the `WindsorContainer.BeginScope` extension. It will also dispose the scope once the request is completed.
+This method also adds a sub resolver for dealing with the resolution of ASP.NET Core framework types. An example might be something like an 
+[ILoggerFactory](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.iloggerfactory?view=aspnetcore-2.0). 
+It is important to note that this extension also injects custom middleware for the management of scoped lifestyles. This middleware calls the 
+`WindsorContainer.BeginScope` extension for scoped lifestyles. It will also dispose the scope once the request is completed.
 
-`Controllers`, `ViewComponents` and `TagHelpers` are registered automatically for you when the `app.UseCastleWindsor<Startup>(container)` extension
-is called from the `Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)` method in the `Startup` class.
-Controllers are registered with a scoped lifestyle, and are assumed to be a single instance for the duration of the web request. The `ViewComponents`
-and `TagHelpers` however are transient. It is assumed that they will called multiple times for a single request and cannot share state.
-If your controllers, view components or taghelpers are in a different assembly, you call this extension multiple times supplying an example
-type for which Windsor will then scan that assembly for framework types.
+### Registering Controllers, ViewComponents and TagHelpers
 
-If you would like to change the lifestyles of any of these components, then you do not call the `app.UseCastleWindsor<Startup>(container)`
-extension. You can opt for your own conventional registrations instead like so:
+`Controllers`, `ViewComponents` and `TagHelpers` are registered automatically for you when the `services.AddCastleWindsor(container)` extension
+is called. All components are registered with a scoped lifestyle, and are assumed to be a single instance for the duration of the web request. 
+You can optionally override the lifestyles for each of these if you prefer using the `options` callback in the example below:
 
 ```csharp
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<Controller>().LifestyleScoped()); // <- Change lifestyle
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<ViewComponent>().LifestyleTransient());
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<TagHelper>().LifestyleTransient());
+services.AddCastleWindsor(container, opts =>
+{
+	opts.RegisterControllers(typeof(HomeController).Assembly, LifestyleType.Transient);
+	opts.RegisterTagHelpers(typeof(EmailTagHelper).Assembly, LifestyleType.Transient);
+	opts.RegisterViewComponents(typeof(AddressComponent).Assembly, LifestyleType.Transient);
```

Comment:
I'm sure you thought about it or used a pattern from somewhere, but could you explain why the class that you perform actions on is named "options".

Reply:
Yeah, this is what it means from an HTTPS point of view.

```
public static void Main(string[] args)
{
	BuildWebHost(args).Run();
}

public static IWebHost BuildWebHost(string[] args) =>
WebHost.CreateDefaultBuilder(args)
	.UseStartup<Startup>()
	.UseKestrel(options =>
	{
		options.Listen(IPAddress.Loopback, 5000);
		options.Listen(IPAddress.Loopback, 5001, listenOptions =>
	{
		listenOptions.UseHttps("testCert.pfx", "testPassword");
	});
	}).Build();
```

Reply:
It is a convention.

Reply:
We can clean this up to make options more stateful but it would mean adding more code.

Reply:
I'll leave it up to you to choose but it just felt a little strange that it was actually doing something rather than just holding info on what the user wanted until calling `AddCastleWindsor` just like like the above does start listening until calling `UseKestrel` or actually probably in `WebHost.Run`.

Couldn't you change `AddApplicationComponentsToWindsor` so rather than looking at boolean members like `ControllersAlreadyRegistered`, have members typed `List<Pair<Assembly, LifeStyleType>>`? It should be about the same amount of code and that way it has the register code once inside `AddApplicationComponentsToWindsor` and it'll either do the stored list or use `options.EntryAssembly`.

Reply:
Do you know what, this was not that bad. I added it in. It did bloat the `AddApplicationComponentsToWindsor` a little but not as bad as I originally thought.

Reply:
Done